modules/core/src/utils/range.js

```diff
@@ -0,0 +1,45 @@
+/* 
+ * range (Array)
+ *   + start (Number) - the start index (incl.)
+ *   + end (Number) - the end index (excl.)
+ * rangeCollection (Array) - array of sorted, combined ranges
+ */
+export const EMPTY = [];
+export const FULL = [[0, Infinity]];
+
+// Insert a range into a range collection
+export function add(rangeCollection, range) {
+  // Noop if range collection already covers all
+  if (rangeCollection === FULL) {
+    return rangeCollection;
+  }
+
+  // Validate the input range
+  if (range[0] < 0) {
+    range[0] = 0;
+  }
+  if (range[0] >= range[1]) {
+    return rangeCollection;
+  }
+
+  const newRangeCollection = [];
```

Comment:
Total nit: Not sure about the "collection" term. Has other connotations to me (base class of containers). I assume you want to emphasize that it has some additional semantics over a list. I think "rangeList" would have been fine.